### ISSUE TITLE
False positives on eta reduction

### ISSUE THREAD

**Issue by llogiq** [Manishearth/rust-clippy#116]
I just managed to run clippy on itself for the first time (using `cargo rustc -- -L ./target/release -Z extra-plugins=clippy`), and tried to quell any warnings that appeared. However, when I fixed eta_reduction warnings on `eq_op.rs`, I ran into errors. For example, it warned:
```
src/eq_op.rs:39:19: 39:44 warning: Redundant closure found, consider using `is_exp_equal` in its place, #[warn(redundant_closure)] on by default
src/eq_op.rs:39 			both(lpl, rpl, |l, r| is_exp_equal(l, r)) && 
                			               ^~~~~~~~~~~~~~~~~~~~~~~~~
```
So I changed it to `both(lpl, rpl, is_exp_equal)`. But after compiling I get:

```
src/eq_op.rs:39:4: 39:8 error: type mismatch: the type `fn(&syntax::ast::Expr, &syntax::ast::Expr) -> bool {eq_op::is_exp_equal}` implements the trait `for<'r, 'r> core::ops::FnOnce<(&'r syntax::ast::Expr, &'r syntax::ast::Expr)>`, but the trait `for<'r, 'r> core::ops::FnOnce<(&'r syntax::ptr::P<syntax::ast::Expr>, &'r syntax::ptr::P<syntax::ast::Expr>)>` is required (expected struct `syntax::ptr::P`, found struct `syntax::ast::Expr`) [E0281]
src/eq_op.rs:39 			both(lpl, rpl, is_exp_equal) && is_exp_equal(lbox, rbox),
                			^~~~
```

It appears the closure is auto-dereferencing the P<Expr>, which overcomes the type mismatch.

**Comment by Manishearth** [comment]
I guess we should use `expr_ty_adjusted` to ensure that there aren't any autoderefs happening; and bail if there is a mismatch between the arg type and arg adjusted type?

**Comment by llogiq** [comment]
Looks like it.

**Comment by llogiq** [comment]
Thinking about it, it seems a bit excessive to call `expr_ty_adjusted` when all we want to know is *if* there is an `AutoAdjustment`.

How about `ctxt.tables.borrow().adjustments.get(&expr.id).is_some()`? Or am I missing something?

**Comment by Manishearth** [comment]
Might work. Never used that part of the code before.

**Comment by llogiq** [comment]
Having tried this, the problem does *not* lie within the closure, but in the `both` function – with a closure it somehow works with a Vec or slice, but with the plain function, it doesn't. See `eq_op.rs`, line 177 – if you try to remove the closure, you get:

```
type mismatch: the type `fn(&[syntax::ptr::P<syntax::ast::Pat>], &[syntax::ptr::P<syntax::ast::Pat>]) -> bool {eq_op::is_pats_equal}` implements the trait `for<'r, 'r> core::ops::FnMut<(&'r [syntax::ptr::P<syntax::ast::Pat>], &'r [syntax::ptr::P<syntax::ast::Pat>])>`, but the trait `for<'r, 'r> core::ops::FnMut<(&'r collections::vec::Vec<syntax::ptr::P<syntax::ast::Pat>>, &'r collections::vec::Vec<syntax::ptr::P<syntax::ast::Pat>>)>` is required (expected struct `collections::vec::Vec`, found slice) [E0281]
src/eq_op.rs:177 		is_path_equal(lpath, rpath) && both(lenum, renum, is_pats_equal),
                 		                               ^~~~
src/eq_op.rs:177:34: 177:38 note: required by `eq_op::both`
src/eq_op.rs:177 		is_path_equal(lpath, rpath) && both(lenum, renum, is_pats_equal),
                 		                               ^~~~
```
It may be that we need some outer context to catch those false positives, but at this point I'm not even sure I understand the problem.

**Comment by llogiq** [comment]
Coming back to this problem after having written the `utils::get_parent` function, we could use it to check if we give the closure as an argument to the function. *Then* we have the Call (or MethodCall) to check if there is a mismatch between the types.

**Comment by llogiq** [comment]
Turns out it is *even harder than I thought*. To match the types we must visit the called fn or method and search for the place(s) where the closure is called. Only then we can match the argument types with the closure's.

**Comment by llogiq** [comment]
Not all is lost. There is one thing in our favor: Closure types are singleton – each closure is its own type. So *if* we can get the method that actually calls the closure, we can visit its expression to see where the closure is called by using expr_ty and comparing with the closure type. At least it doesn't require symbolic execution.

**Comment by llogiq** [comment]
Given that the function very likely does not even know the closure type, I'm currently leaning to the very crude approximation of quelling the lint whenever we find any `ExprCall` or `ExprMethodCall` where any argument has the same `expr_ty` as the closure.

Perhaps we could store both offending closures and all closure types from call arguments and only report after having visited everything? Is there even a callback for that?

**Comment by llogiq** [comment]
I'm leaving this for now – the crude approximation would effectively disable the lint (unless you count our very synthetic test case). In an ideal world, Rust would implement Deref coercions for both closures and method pointer calls. As it stands, we have next to no options of making this work.

**Comment by Manishearth** [comment]
I'm still skeptical that we can't bypass this with adjustments

**Comment by Manishearth** [comment]
But sure

**Comment by llogiq** [comment]
We'd still have two problems: The *absence check* (check there *is no* call with adjusted types) requires checking all possible references to the closure – if the closure is a public item, we've lost anyway, and the method check – we'd need to walk the whole call tree (or at least the part of it where the closure is taken to) and check each (Method)Call with `expr_ty`. There, we'll have to somehow unify the types with all generics the function may introduce, and by now we're looking at a serious headache.

**Comment by Manishearth** [comment]
Okay, so what if we only lint closures being used immediately? where the parent node of the closure is an `Expr*Call`

**Comment by Manishearth** [comment]
(I'm not yet clear on what the problem is here)

**Comment by llogiq** [comment]
Ok, that solves the first problem. We still have the second. As a simple example, try and implement a factorial function using a generic Y combinator.

**Comment by Manishearth** [comment]
I don't see why we need to do (2). The adjustments should be applied at the original call site, not the method tree.

**Comment by llogiq** [comment]
Ok then. I'll see what I can do.
